src/lib/support/BytesCircularBuffer.cpp

```diff
@@ -0,0 +1,165 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <support/BytesCircularBuffer.h>
+
+#include <limits>
+#include <nlassert.h>
+#include <string.h>
+
+#include <support/CodeUtils.h>
+
+namespace chip {
+
+size_t BytesCircularBuffer::Advance(size_t dataLocation, size_t amount) const
+{
+    dataLocation += amount;
+    if (dataLocation >= mCapacity)
+        dataLocation -= mCapacity;
+    return dataLocation;
+}
+
+void BytesCircularBuffer::Read(uint8_t * dest, size_t length, size_t offset) const
+{
+    // length is an instance of SizeType
+    // offset is maximized at sizeof(SizeType) for all use cases.
+    static_assert(std::numeric_limits<SizeType>::max() < std::numeric_limits<size_t>::max() - sizeof(SizeType),
+                  "SizeType too large, may cause overflow");
+    VerifyOrDie(StorageUsed() >= offset + length);
+
+    size_t start = Advance(mDataStart, offset);
+    if (mCapacity - start >= length)
+    {
+        ::memcpy(dest, mStorage + start, length);
+    }
+    else
+    {
+        size_t firstPiece  = mCapacity - start;
+        size_t secondPiece = length - firstPiece;
+        ::memcpy(dest, mStorage + start, firstPiece);
+        ::memcpy(dest + firstPiece, mStorage, secondPiece);
+    }
+}
+
+void BytesCircularBuffer::Write(const uint8_t * source, size_t length)
+{
+    // Always reserve 1 byte to prevent mDataStart == mDataEnd because then it would be
+    // ambiguous whether we have 0 bytes or mCapacity bytes stored.
+    VerifyOrDie(mCapacity - StorageUsed() - 1 >= length);
+
+    if (mCapacity - mDataEnd >= length)
+    {
+        ::memcpy(mStorage + mDataEnd, source, length);
+        mDataEnd = Advance(mDataEnd, length); // Use advance in case that mDataEnd wrap to 0
+    }
+    else
+    {
+        size_t firstPiece  = mCapacity - mDataEnd;
```

Comment:
Same here, I don't think the condition is needed.

- Use std::min() here.
- Assign mDataEnd as in the 1st case: `mDataEnd = Advance(mDataEnd, length);`

memcpy of zero bytes is fine.